File: Prob23.java
import java.util.List;
import java.util.ArrayList;

class Prob23 {
    
    private List<Integer> abundants = new ArrayList<Integer>();

    public static final int MAX = 28123;

    public Prob23() {
	
    }

    private int sum(List<Integer> lst) {
	int sum = 0;
	for (Integer i : lst) {
	    sum += i;
	}
	return sum;
    }
    
    private List<Integer> devisors(int num) {
	List<Integer> divisors = new ArrayList<Integer>();
	for (int i = 1; i < (num/2+1); i++) {
	    divisors.add(i);
	}
	return divisors;
    }

    private int abundantNum(int num) {
	if (num == 2) {
	    return 1;
	}
	int sum = 0;
	for (int i = 1; i < (num/2+1); i++) {
	    if (num % i == 0) {
		sum += i;
	    }
	}
	return sum;
    }
    
    private boolean isAbundant(int num) {
	return num < abundantNum(num);
    }

    public int answer() {
	List<Integer> abundants = new ArrayList<Integer>();
	for (int i = 1; i < MAX; i++) {
	    if (!isAbundant(i)) {
		continue;
	    }
	    abundants.add(i);
	}
	System.out.println(abundants.size());
	List<Integer> sumofabundants = new ArrayList<Integer>();
	for (int i : abundants) {
	    for (int j : abundants) {
		if ((i+j) < MAX && !sumofabundants.contains(i+j)) {
		    sumofabundants.add(i+j);
		}
	    }
	    sumofabundants.remove(i);
	}
	System.out.println(sumofabundants.size());
	int sum = 0;
	for (int i = 1; i < MAX; i++) {
	    if (!sumofabundants.contains(i)) {
		sum += i;
	    }
	}
	return sum;
    }

    public static void main(String[] args) {
	Prob23 p = new Prob23();
	System.out.println(p.answer());
    }
}

File: prob23.rb
MAX = 28123

class Prob23

  MAX = 28123

  @seed = 1..MAX

  def initialize
  end

  def divisors(num)
    divisors = Array.new
    for i in 1..(num/2+1)
      divisors << i if num % i == 0
    end
    return divisors
  end

  def sum(list)
    sum = 0
    list.map{|v|
      sum += v
    }
    return sum
  end

  def abundantNum(num)
    return 1 if num == 2
    abundantNum = 0
    for i in 1..(num/2+1)
      abundantNum += i if num % i == 0
    end
    return abundantNum
  end

  def isAbundant?(num)
    return num < abundantNum(num)
  end

  def main
    start = Time.new
    print start, "\n"
    @abundants = []
    for i in 1..MAX
      next if !isAbundant?(i)
      @abundants << i
    end
    print "abundants is solved.(", Time.new - start, ")\n"
    @sumofabundants = []
    @abundants.each{|i|
      @abundants.each{|j|
        @sumofabundants << i+j if i+j < MAX and @sumofabundants.index(i+j) == nil
      }
      @abundants.pop
    }
    print "sumofabundants is solved.(", Time.new - start, ")\n"
    sum = 0
    for i in 1..MAX
      sum += i if @sumofabundants.index(i) == nil
    end
    print sum
    print Time.new - start
  end
end

main = Prob23.new
main.main
